src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts

```diff
@@ -0,0 +1,327 @@
+  isEmpty$ = this.store$.select(selectSnapshotsTotal).pipe(map((total) => total === 0));
+  @ViewChild(MatSort, { static: false }) sort: MatSort;
+  dataSource: MatTableDataSource<SnapshotListRow> = new MatTableDataSource([]);
+  defaultSort: Sort = { active: 'snapshot', direction: 'desc' };
+  filterString = '';
+  expandedRow: SnapshotListRow;
+  selection = new SelectionModel(true, []);
+  settingsEvent$: Subject<CoreEvent> = new Subject();
+  showExtraColumns$ = new BehaviorSubject(false);
+  toolbarConfig: ToolbarConfig;
+  emptyConfig: EmptyConfig = {
+    type: EmptyType.NoPageData,
+    title: this.translate.instant('No snapshots are available.'),
+    large: true,
+  };
+  loadingConfig: EmptyConfig = {
+    type: EmptyType.Loading,
+    large: false,
+    title: this.translate.instant('Loading...'),
+  };
+  errorConfig: EmptyConfig = {
+    type: EmptyType.Errors,
+    large: true,
+    title: this.translate.instant('Can not retrieve response'),
+  };
+
+  get displayedColumns(): string[] {
+    if (this.showExtraColumns$.value) {
+      return ['select', 'dataset', 'snapshot', 'used', 'created', 'referenced', 'actions'];
+    }
+    return ['select', 'dataset', 'snapshot', 'actions'];
+  }
+
+  constructor(
+    private dialogService: DialogService,
+    private ws: WebSocketService,
+    private translate: TranslateService,
+    private modalService: ModalService,
+    private cdr: ChangeDetectorRef,
+    private core: CoreService,
+    private loader: AppLoaderService,
+    private matDialog: MatDialog,
+    private store$: Store<SnapshotSlice>,
+  ) {
+    if (window.localStorage.getItem('snapshotXtraCols') === 'true') {
+      this.showExtraColumns$.next(true);
+    }
+  }
+
+  ngOnInit(): void {
+    this.setupToolbar();
+    this.getSnapshots();
+  }
+
+  ngAfterViewInit(): void {
+    this.store$.dispatch(loadSnapshots({ extra: this.showExtraColumns$.value }));
+    this.modalService.onClose$.pipe(untilDestroyed(this)).subscribe(() => {
+      this.getSnapshots();
+    });
+  }
+
+  getSnapshots(): void {
+    this.store$.select(selectSnapshots).pipe(
+      map((snapshots: ZfsSnapshot[]) => {
+        const snapshotListRows = [];
+
+        for (const row of snapshots) {
+          const [datasetName, snapshotName] = row.name.split('@');
+
+          const transformedRow = {
+            id: row.name,
+            dataset: datasetName,
+            snapshot: snapshotName,
+            properties: row.properties,
+            name: row.name,
+          } as SnapshotListRow;
+
+          if (row.properties) {
+            snapshotListRows.push({
+              ...transformedRow,
+              used: parseInt(row.properties.used.rawvalue),
+              created: row.properties.creation.parsed.$date,
+              referenced: parseInt(row.properties.referenced.rawvalue),
+            });
+          } else {
+            snapshotListRows.push(transformedRow);
+          }
+        }
+
+        return snapshotListRows;
```

Comment:
This processing belongs to a selector (or alternatively you can avoid transforming at all and just read necessary values in a template).